aws/rust-runtime/aws-config/src/standard_property.rs

```diff
@@ -94,28 +125,238 @@ impl StandardProperty {
         &'a self,
         provider_config: &'a ProviderConfig,
     ) -> Option<(Cow<'a, str>, PropertySource<'a>)> {
-        if let Some(env_var) = self.environment_variable.as_ref() {
-            if let Ok(value) = provider_config.env().get(env_var) {
-                return Some((
-                    Cow::Owned(value),
-                    PropertySource::Environment { name: env_var },
-                ));
-            }
-        }
-        if let Some(profile_key) = self.profile_key.as_ref() {
-            let profile = provider_config.profile().await?;
-
-            if let Some(value) = profile.get(profile_key) {
-                return Some((
-                    Cow::Borrowed(value),
-                    PropertySource::Profile {
-                        name: profile.selected_profile(),
-                        key: profile_key,
-                    },
-                ));
-            }
-        }
+        let env_value = self.environment_variable.as_ref().and_then(|env_var| {
+            // Check for a service-specific env var first
+            get_service_config_from_env(provider_config, self.service_id.as_ref(), env_var)
+                // Then check for a global env var
+                .or_else(|| {
+                    provider_config.env().get(env_var).ok().map(|value| {
+                        (
+                            Cow::Owned(value),
+                            PropertySource::Environment {
+                                name: env_var,
+                                service_specific: false,
+                            },
+                        )
+                    })
+                })
+        });
+
+        let profile = provider_config.profile().await?;
+        let profile_value = self.profile_key.as_ref().and_then(|profile_key| {
+            // Check for a service-specific profile key first
+            get_service_config_from_profile(profile, self.service_id.as_ref(), profile_key)
+                // Then check for a global profile key
+                .or_else(|| {
+                    profile.get(profile_key).map(|value| {
+                        (
+                            Cow::Borrowed(value),
+                            PropertySource::Profile {
+                                name: profile.selected_profile(),
+                                key: profile_key,
+                                service_specific: false,
+                            },
+                        )
+                    })
+                })
+        });
+
```

Comment:
there's probably some opportunity to cleanup / unify this code a little bit but no worries